### DISCUSSION TITLE
is it not possible to use vite's mode option for vitepress build?

### DISCUSSION THREAD

**Discussion by linusang-spark** [vuejs/vitepress#3533] [Q&A]
it seems that the mode is only `development` when we run `docs:dev` and `production` when we run `docs:build`. No other custom modes we can use?

**Comment by brc-dd** [comment]
Yeah, not supported.

**Reply by linusang-spark** [reply]
ok that's unfortunate :(

**Reply by brc-dd** [reply]
What’s your use case?

**Reply by linusang-spark** [reply]
i am loading some dynamic data on the vitepress site and we want to test it under a `preprod` environment before publishing it. I have a `.env.preprod` file that has a different environment value from the  `.env.production` file. I foresee there might be use-cases for a staging or testing env where vite's `mode` option will be useful. would be great if vitepress can support this flag for `vitepress build`

**Reply by brc-dd** [reply]
You can do something like this:

```jsonc
// package.json scripts:

"docs:build:staging": "VP_MODE=staging vitepress build docs"
```

```ts
// docs/.vitepress/config.ts

import { loadEnv, defineConfig } from 'vitepress'

loadEnv(process.env.VP_MODE || '', process.cwd())

export default defineConfig(...)
```

On newer versions of node, you can just do something like:

```jsonc
"docs:build:staging": "node --env-file=.env.staging ./node_modules/vitepress/bin/vitepress.js"
```

**Reply by linusang-spark** [reply]
thanks for the solution. On a similar question, this wouldn't work on `<ClientOnly>` components right? Because when I tried, it is either development or production mode it seems.

**Reply by SetsuikiHyoryu** [reply]
> You can do something like this:
> 
> ```js
> // package.json scripts:
> 
> "docs:build:staging": "VP_MODE=staging vitepress build docs"
> ```
> 
> ```ts
> // docs/.vitepress/config.ts
> 
> import { loadEnv, defineConfig } from 'vitepress'
> 
> loadEnv(process.env.VP_MODE || '', process.cwd())
> 
> export default defineConfig(...)
> ```
> 
> On newer versions of node, you can just do something like:
> 
> ```js
> "docs:build:staging": "node --env-file=.env.staging ./node_modules/vitepress/bin/vitepress.js"
> ```

I tried this method in `docs/.vitepress/config.ts`. I followed the approach in the [official documentation](https://vitepress.dev/reference/site-config#config-resolution) to dynamically return defineConfig using a function, and I was surprised to find that this function receives the same parameters as Vite, allowing me to successfully access environment variables.

```ts
// docs/.vitepress/config.ts
export default async ({ mode }) => {
  const env = loadEnv(mode || '', process.cwd())
  retrun defineConfig({ /* ... */ })
}
```

**Comment by rwam** [comment]
I've tried a workaround like described using `VP_MODE` to use different variables on a staging environment but with now luck. It's not critical for us so we have to go with this kind of limitation. But it's annoying, sorry. Why cannot simply Vitepress support vite's `--mode` option flag resp. [modes](https://vite.dev/guide/env-and-mode#modes)?
